Refactor Gaussian Likelihoods lnValue Interface

pbauman opened the pull request:
This is step 1 for refactoring/adding marginalization to the (Gaussian) likelihoods. First, renamed the base class to `LikelihoodBase` because there's nothing Gaussian about the base class. Then moved `actualValue` to the base class since all the subclasses were exactly the same function. Finally, moved to the preferred `lnValue(domainVector)` interface (from #493); along those lines, added a new `evaluateModel` interface that takes only the `domainVector` and the `modelOutput`.

pbauman commented:
I just tested this on a small QUESO app which uses `GaussianLikelihoodDiagonalCovariance` and everything worked as before, so I think the backward compatibility is good.

dmcdougall commented:
I get a compiler error on this branch.  Here're my loaded modules:

```
$ module list

Currently Loaded Modules:
  1) c7   2) autotools/1.2   3) gdb/7.10   4) gcc/5.2   5) gsl/2.0   6) mpich2/3.1.4   7) texlive/2015-11-05   8) cppunit/1.10.3   9) boost/1.58.0
```

And here's my `configure` line:

```
../configure --prefix=$HOME/queso
```

And here's the output from `make`:

```
make  all-recursive
make[1]: Entering directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods'
Making all in inc/queso
make[2]: Entering directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/inc/queso'
make  all-am
make[3]: Entering directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/inc/queso'
make[3]: Nothing to be done for `all-am'.
make[3]: Leaving directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/inc/queso'
make[2]: Leaving directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/inc/queso'
Making all in src/contrib/ANN
make[2]: Entering directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/src/contrib/ANN'
make[2]: Nothing to be done for `all'.
make[2]: Leaving directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/src/contrib/ANN'
Making all in src
make[2]: Entering directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/src'
make  all-am
make[3]: Entering directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/src'
  CXX      stats/src/LikelihoodBase.lo
../../src/stats/src/LikelihoodBase.C: In member function ‘virtual void QUESO::LikelihoodBase<V, M>::evaluateModel(const V&, const V*, V&, V*, M*, V*) const’:
../../src/stats/src/LikelihoodBase.C:52:6: error: no match for ‘operator<<’ (operand types are ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ and ‘const char [70]’)
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
      ^
In file included from ../inc/queso/Vector.h:28:0,
                 from ../inc/queso/GslVector.h:34,
                 from ../../src/stats/src/LikelihoodBase.C:25:
../inc/queso/Environment.h:492:15: note: candidate: std::ostream& QUESO::operator<<(std::ostream&, const QUESO::BaseEnvironment&)
 std::ostream& operator<<(std::ostream& os, const BaseEnvironment& obj);
               ^
../inc/queso/Environment.h:492:15: note:   no known conversion for argument 1 from ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ to ‘std::ostream& {aka std::basic_ostream<char>&}’
In file included from ../../src/stats/src/LikelihoodBase.C:25:0:
../inc/queso/GslVector.h:258:18: note: candidate: std::ostream& QUESO::operator<<(std::ostream&, const QUESO::GslVector&)
 std::ostream&    operator<<   (std::ostream& os,            const GslVector& obj);
                  ^
../inc/queso/GslVector.h:258:18: note:   no known conversion for argument 1 from ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ to ‘std::ostream& {aka std::basic_ostream<char>&}’
In file included from ../../src/stats/src/LikelihoodBase.C:26:0:
../inc/queso/GslMatrix.h:479:18: note: candidate: std::ostream& QUESO::operator<<(std::ostream&, const QUESO::GslMatrix&)
 std::ostream&    operator<<      (std::ostream& os,            const GslMatrix& obj);
                  ^
../inc/queso/GslMatrix.h:479:18: note:   no known conversion for argument 1 from ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ to ‘std::ostream& {aka std::basic_ostream<char>&}’
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/memory:82:0,
                 from ../inc/queso/ScopedPtr.h:31,
                 from ../inc/queso/Environment.h:32,
                 from ../inc/queso/Vector.h:28,
                 from ../inc/queso/GslVector.h:34,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/bits/shared_ptr.h:66:5: note: candidate: template<class _Ch, class _Tr, class _Tp, __gnu_cxx::_Lock_policy _Lp> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, const std::__shared_ptr<_Tp, _Lp>&)
     operator<<(std::basic_ostream<_Ch, _Tr>& __os,
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/bits/shared_ptr.h:66:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/asserts.h:40:0,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:363:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, std::_Put_time<_CharT>)
     operator<<(basic_ostream<_CharT, _Traits>& __os, _Put_time<_CharT> __f)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:363:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/asserts.h:40:0,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:311:5: note: candidate: template<class _CharT, class _Traits, class _MoneyT> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, std::_Put_money<_MoneyT>)
     operator<<(basic_ostream<_CharT, _Traits>& __os, _Put_money<_MoneyT> __f)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:311:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/asserts.h:40:0,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:238:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, std::_Setw)
     operator<<(basic_ostream<_CharT, _Traits>& __os, _Setw __f)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:238:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/asserts.h:40:0,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:208:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, std::_Setprecision)
     operator<<(basic_ostream<_CharT, _Traits>& __os, _Setprecision __f)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:208:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/asserts.h:40:0,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:178:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, std::_Setfill<_CharT>)
     operator<<(basic_ostream<_CharT, _Traits>& __os, _Setfill<_CharT> __f)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:178:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/asserts.h:40:0,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:143:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, std::_Setbase)
     operator<<(basic_ostream<_CharT, _Traits>& __os, _Setbase __f)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:143:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/asserts.h:40:0,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:109:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, std::_Setiosflags)
     operator<<(basic_ostream<_CharT, _Traits>& __os, _Setiosflags __f)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:109:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/asserts.h:40:0,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:79:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, std::_Resetiosflags)
     operator<<(basic_ostream<_CharT, _Traits>& __os, _Resetiosflags __f)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iomanip:79:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:628:5: note: candidate: template<class _CharT, class _Traits, class _Tp> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&&, const _Tp&)
     operator<<(basic_ostream<_CharT, _Traits>&& __os, const _Tp& __x)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:628:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:574:5: note: candidate: template<class _Traits> std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, const unsigned char*)
     operator<<(basic_ostream<char, _Traits>& __out, const unsigned char* __s)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:574:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<char, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:569:5: note: candidate: template<class _Traits> std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, const signed char*)
     operator<<(basic_ostream<char, _Traits>& __out, const signed char* __s)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:569:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<char, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:556:5: note: candidate: template<class _Traits> std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, const char*)
     operator<<(basic_ostream<char, _Traits>& __out, const char* __s)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:556:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<char, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:638:0,
                 from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/bits/ostream.tcc:321:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, const char*)
     operator<<(basic_ostream<_CharT, _Traits>& __out, const char* __s)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/bits/ostream.tcc:321:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:539:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, const _CharT*)
     operator<<(basic_ostream<_CharT, _Traits>& __out, const _CharT* __s)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:539:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:519:5: note: candidate: template<class _Traits> std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, unsigned char)
     operator<<(basic_ostream<char, _Traits>& __out, unsigned char __c)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:519:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<char, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:514:5: note: candidate: template<class _Traits> std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, signed char)
     operator<<(basic_ostream<char, _Traits>& __out, signed char __c)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:514:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<char, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:508:5: note: candidate: template<class _Traits> std::basic_ostream<char, _Traits>& std::operator<<(std::basic_ostream<char, _Traits>&, char)
     operator<<(basic_ostream<char, _Traits>& __out, char __c)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:508:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<char, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:502:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, char)
     operator<<(basic_ostream<_CharT, _Traits>& __out, char __c)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:502:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39:0,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:497:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, _CharT)
     operator<<(basic_ostream<_CharT, _Traits>& __out, _CharT __c)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:497:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/bits/ios_base.h:46:0,
                 from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ios:42,
                 from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/ostream:38,
                 from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iostream:39,
                 from ../inc/queso/asserts.h:39,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/system_error:209:5: note: candidate: template<class _CharT, class _Traits> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, const std::error_code&)
     operator<<(basic_ostream<_CharT, _Traits>& __os, const error_code& __e)
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/system_error:209:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/string:52:0,
                 from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/stdexcept:39,
                 from ../inc/queso/exceptions.h:29,
                 from ../inc/queso/asserts.h:35,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/bits/basic_string.h:5170:5: note: candidate: template<class _CharT, class _Traits, class _Alloc> std::basic_ostream<_CharT, _Traits>& std::operator<<(std::basic_ostream<_CharT, _Traits>&, const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&)
     operator<<(basic_ostream<_CharT, _Traits>& __os,
     ^
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/bits/basic_string.h:5170:5: note:   template argument deduction/substitution failed:
../../src/stats/src/LikelihoodBase.C:52:9: note:   ‘std::stringstream {aka std::__cxx11::basic_stringstream<char>}’ is not derived from ‘std::basic_ostream<_CharT, _Traits>’
   ss << "ERROR: evaluateModel() not implemented! This interface is deprecated."
         ^
In file included from ../inc/queso/Defines.h:29:0,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
../../src/stats/src/LikelihoodBase.C:57:21: error: invalid use of incomplete type ‘std::stringstream {aka class std::__cxx11::basic_stringstream<char>}’
   queso_error_msg(ss.str());
                     ^
../inc/queso/asserts.h:47:62: note: in definition of macro ‘queso_error_msg’
 #define queso_error_msg(msg) do { queso_here(); std::cerr << msg << std::endl; QUESO_THROW(QUESO::LogicError()); } while(0)
                                                              ^
In file included from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/bits/localefwd.h:41:0,
                 from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/string:43,
                 from /opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/stdexcept:39,
                 from ../inc/queso/exceptions.h:29,
                 from ../inc/queso/asserts.h:35,
                 from ../inc/queso/Defines.h:29,
                 from ../inc/queso/GslVector.h:33,
                 from ../../src/stats/src/LikelihoodBase.C:25:
/opt/apps/ossw/applications/gcc/gcc-5.2/c7/include/c++/5.2.0/iosfwd:108:11: note: declaration of ‘std::stringstream {aka class std::__cxx11::basic_stringstream<char>}’
     class basic_stringstream;
           ^
../../src/stats/src/LikelihoodBase.C: In instantiation of ‘void QUESO::LikelihoodBase<V, M>::evaluateModel(const V&, const V*, V&, V*, M*, V*) const [with V = QUESO::GslVector; M = QUESO::GslMatrix]’:
../../src/stats/src/LikelihoodBase.C:74:23:   required from here
../../src/stats/src/LikelihoodBase.C:51:21: error: ‘std::stringstream ss’ has incomplete type
   std::stringstream ss;
                     ^
make[3]: *** [stats/src/LikelihoodBase.lo] Error 1
make[3]: Leaving directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/src'
make[2]: *** [all] Error 2
make[2]: Leaving directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods/src'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/scratch/damon/git/queso/build_refactor-gaussian-likelihoods'
make: *** [all] Error 2
```

pbauman commented:
Weird. Neither Travis (I think GCC 4.7?) and mine (GCC 6.3) got this. Wonder if there's some header shenanigans in the STL in GCC 5.2. I'll try and confirm this failure with a GCC 5.2 build locally.

pbauman commented:
Ok, I had 5.3 around and I can reproduce

pbauman commented:
OK, pushed a fix that gets me compiling with GCC 5.3.

pbauman commented:
Went ahead and rebased on dev while I was at it.

roystgnr commented:
That's dumbfounding!

"Missing header" was my first guess, but then I saw ```std::stringstream {aka std::__cxx11::basic_stringstream<char>``` which didn't seem to think stringstream was an incomplete type ... but I guess that was just a forward declaration, and it's the basic_stringstream declaration that was missing?

pbauman commented:
Yeah, I don't know. I tried Googling for a couple of mins. but couldn't hit anything.

dmcdougall commented:
You're welcome.

pbauman commented:
Rebased on dev after #532 was merged.

dmcdougall commented on src/stats/inc/LikelihoodBase.h:
```diff
@@ -45,7 +46,7 @@ class GslMatrix;
 
 template <class V = GslVector, class M = GslMatrix>
-class BaseGaussianLikelihood : public BaseScalarFunction<V, M> {
+class LikelihoodBase : public BaseScalarFunction<V, M> {
```
We're operating under the assumption users wanting Gaussian likelihoods would have subclassed one of the existing Gaussian likelihoods to handle their covariance case rather than subclassing `BaseGaussianLikelihood` to *extend* QUESO's existing Gaussian likelihood capabilities.

pbauman replied:
That is correct. My thinking is the user may want to use this interface for some other likelihood function that's not Gaussian. Thus the name would be more consistent, IMHO.

dmcdougall commented on src/stats/inc/LikelihoodBase.h:
```diff
@@ -77,10 +85,31 @@ class BaseGaussianLikelihood : public BaseScalarFunction<V, M> {
+                                  hessianMatrix, hessianEffect)); }
+
+  //! Logarithm of the value of the scalar function.
+  virtual double lnValue(const V & domainVector) const =0;
```
Can you please explain why this class is re-declaring these two `lnValue` methods?  Moreover,

1.  This class re-declares a virtual but non-pure `virtual BaseScalarFunction::lnValue(const V &) const` into the pure virtual `virtual LikelihoodBase::lnValue(const V &) const = 0`;
2.  This class re-declares `lnValue(const V &, const V *, V *, M *, V *) const`, and implements it.  The implementation calls `lnValue(const V &) const`.  Can you explain why?  In the base class `BaseScalarFunction` this logic is exactly reversed; that is to say, `lnValue(const V &, const V *, V *, M *, V *) const` calls `lnValue(const V &) const`.

pbauman replied:
1. Oops, copy/paste error, I'll fix it (remove the pure part). I still like to have the virtual keyword there to clue us in that this may be overriding a function. Thanks for catching this.

2. Because I was thinking about `evaluateModel` backward compatibility and mistakenly applied that logic to `lnValue` as well. Thank you for pointing this out.

dmcdougall replied:
Gotcha.  I spent a little while trying to figure out what the reasoning behind it was before I gave up and just resorted to asking you to explain it to me.  Glad I did :)

pbauman commented:
@dmcdougall OK, so I was going to back to make those fixes and I've noticed, through subsequent test failures, that there are several places internal to QUESO still using the old `lnValue` interface (that is directly call the 5 argument version) - is this expected?

dmcdougall commented:
I don't think so.  Can you point me to an instance so I may investigate further?

pbauman commented:
E.g. https://github.com/libqueso/queso/blob/dev/src/stats/src/BayesianJointPdf.C#L197 (several places in that file)

https://github.com/libqueso/queso/blob/dev/src/basic/src/ScalarFunctionSynchronizer.C#L246

Those popped up from the test failures, but I've only looked at one of them so there may be other places. I can do a pass at these, but I didn't want to put the time in before confirming with you they should be changed.

dmcdougall commented:
Yes, they should be changed.  It isn't a case of just calling the version of `lnValue` that takes only one value, though.  For example the `callFunction` method [here](https://github.com/libqueso/queso/blob/dev/src/basic/src/ScalarFunctionSynchronizer.C#L62) would need overloaded versions that call the right version of `lnValue`.

The reason for this is that QUESO should be calling the version of `lnValue` (or `callFunction`, or whatever) corresponding to the amount of information (gradients) it needs.

pbauman commented:
Cool. In that case, if you know the right fixes in the right places, would this be a quick fix for you? If not, I can tackle it.

pbauman commented:
[queso.txt](https://github.com/libqueso/queso/files/733913/queso.txt)

Here's the patch at the tip of this branch that trips the failures I'm seeing.

dmcdougall commented:
See #537.

pbauman commented:
Rebased, squashed in the changes for consistent usage of `lnValue` and then added commit 36ff3f5 which updates the Gaussian likelihood tests to use the new API. Note that commit 59c335d also passed `make check`, ensuring the backward compatibility was working as expected.

I believe this should be good to go now, unless you have other comments @dmcdougall.
